src/tasks/withdrawService.ts

```diff
@@ -0,0 +1,385 @@
+  usdReference,
+  hre,
+  api,
+  dryRun,
+}: WithdrawAndDumpInput): Promise<State> {
+  // Update list of pending tokens to determine which token was traded
+  const pendingTokens = (
+    await Promise.all(
+      state.pendingTokens.map(async (pendingToken) => {
+        if (pendingToken.retries >= MAX_ORDER_RETRIES_BEFORE_SKIPPING) {
+          // Note that this error might be triggered in legitimate cases, for
+          // example if a token did not trade the first time and then the price
+          // has become so low that it's not enough to pay for the fee.
+          // TODO: revisit after getting an idea of the frequency at which this
+          // alert is triggered.
+          console.error(
+            `Tried ${pendingToken.retries} times to sell token ${pendingToken.address} without success. Skipping token until future run`,
+          );
+          return [];
+        }
+
+        // assumption: eth is not in the list (as it's not supported by the
+        // withdraw script).
+        const token = await erc20Token(pendingToken.address, hre);
+        if (token === null) {
+          throw new Error(
+            `Previously sold a token that is not a valid ERC20 token anymore (address ${pendingToken.address})`,
+          );
+        }
+        pendingToken.retries += 1;
+        return (await balanceOf(token, solver.address)).isZero()
+          ? []
+          : [pendingToken];
+      }),
+    )
+  ).flat();
+
+  console.log("Recovering list of tokens traded since the previous run...");
+  const recentlyTradedTokens = await getAllTradedTokens(
+    settlement,
+    state.lastUpdateBlock,
+    latestBlock,
+    hre,
+  );
+
+  const tradedTokens = state.tradedTokens.concat(
+    recentlyTradedTokens.filter(
+      (token) =>
+        !(state.tradedTokens.includes(token) || token === BUY_ETH_ADDRESS),
+    ),
+  );
+  const numCheckedTokens = Math.min(
+    MAX_CHECKED_TOKENS_PER_RUN,
+    tradedTokens.length,
+  );
+  // The index of the checked token wraps around after reaching the end of the
+  // traded token list
+  const checkedTokens = tradedTokens
+    .concat(tradedTokens)
+    .slice(state.nextTokenToTrade, state.nextTokenToTrade + numCheckedTokens);
+  const updatedState: State = {
+    lastUpdateBlock: latestBlock,
+    tradedTokens,
+    nextTokenToTrade:
+      (state.nextTokenToTrade + numCheckedTokens) % tradedTokens.length,
+    pendingTokens,
+  };
+
+  const withdrawnTokens = await withdraw({
+    solver,
+    tokens: checkedTokens,
+    minValue,
+    leftover,
+    receiver: solver.address,
+    authenticator,
+    settlement,
+    settlementDeploymentBlock,
+    latestBlock,
+    network,
+    usdReference,
+    hre,
+    api,
+    dryRun,
+    doNotPrompt: true,
+  });
+
+  const tokensToDump = Array.from(
+    new Set(pendingTokens.map((t) => t.address).concat(withdrawnTokens)),
+  ).filter((addr) => addr !== BUY_ETH_ADDRESS);
+
+  // wait for node to pick up updated balances
+  await sleep(5000);
```

Comment:
Maybe we can easily set min confirmations to something like 2 blocks so we don't need this probabilistic sleep here (seems risky to assume balances are updated after 5s)?